client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -604,17 +605,34 @@ private Boolean handleScaleResponse(Stream stream, ScaleResponse response) {
         Exceptions.checkNotClosed(closed.get(), this);
         Stream stream = from.getStream();
         long traceId = LoggerHelpers.traceEnter(log, "getSuccessorsFromCut", stream);
-        HashSet<Segment> unread = new HashSet<>(from.getPositions().keySet());
         val currentSegments = getAndHandleExceptions(getCurrentSegments(stream.getScope(), stream.getStreamName()),
-                                                     RuntimeException::new);
-        unread.addAll(computeKnownUnreadSegments(currentSegments, from));   
+                                                     RuntimeException::new).getSegments();
+        Set<Segment> unread = getSegmentsInclusive(from, currentSegments);
+        LoggerHelpers.traceLeave(log, "getSuccessorsFromCut", traceId);
+        return CompletableFuture.completedFuture(unread);
+    }
+
+    @Override
+    public CompletableFuture<Set<Segment>> getSegmentsInclusive(StreamCut fromStreamCut, StreamCut toStreamCut) {
+        Preconditions.checkArgument(fromStreamCut.getStream().equals(toStreamCut.getStream()), "Ensure streamCuts for the same stream is passed");
+        //TODO: Validation check to ensure toStreamCut >= fromStreamCut.
+        Stream stream = fromStreamCut.getStream();
+        long traceId = LoggerHelpers.traceEnter(log, "getSuccessorsFromCut", stream);
+        Set<Segment> segments = getSegmentsInclusive(fromStreamCut, toStreamCut.getPositions().keySet());
+        LoggerHelpers.traceLeave(log, "getSuccessorsFromCut", traceId);
+        return CompletableFuture.completedFuture(segments);
+    }
+
+    private Set<Segment> getSegmentsInclusive(StreamCut fromStreamCut, Collection<Segment> currentSegments) {
+        HashSet<Segment> unread = new HashSet<>(fromStreamCut.getPositions().keySet());
+        unread.addAll(currentSegments);
+        unread.addAll(computeKnownUnreadSegments(currentSegments, fromStreamCut));
         ArrayDeque<Segment> toFetchSuccessors = new ArrayDeque<>();
-        for (Segment toFetch : from.getPositions().keySet()) {
+        for (Segment toFetch : fromStreamCut.getPositions().keySet()) {
             if (!unread.contains(toFetch)) {
                 toFetchSuccessors.add(toFetch);
             }
         }
-        unread.addAll(currentSegments.getSegments());
         while (!toFetchSuccessors.isEmpty()) {
```

Comment:
instead of iterating over toFetchSuccessors, you can initiate multiple futures in parallel and combine them.